pkg/kubectl/util/service_port.go

```diff
@@ -0,0 +1,66 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"fmt"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+// Lookup containerPort number by its named port name
+func LookupContainerPortNumberByName(pod v1.Pod, name string) (int32, error) {
+	matchfound := false
+	namedport := int32(-1)
+	for _, ctr := range pod.Spec.Containers {
+		for _, ctrportspec := range ctr.Ports {
+			if ctrportspec.Name == name {
+				namedport = ctrportspec.ContainerPort
+				matchfound = true
+			}
+		}
+	}
+
+	if matchfound == false {
+		return namedport, fmt.Errorf("Pod '%s' does not have a named port '%s'", pod.Name, name)
+	} else {
+		return namedport, nil
+	}
+}
+
+// Gather service port to target pod mapping from a Service
+// It implements the handling of resolving container named port, as well as ignoring targetPort when clusterIP=None
+func GatherServicePortMapping(svc v1.Service, pod v1.Pod) (map[int32]int32, error) {
+	mapping := make(map[int32]int32)
+	for _, svcportspec := range svc.Spec.Ports {
+		if svc.Spec.ClusterIP == "None" {
+			mapping[svcportspec.Port] = svcportspec.Port
+		} else {
+			if svcportspec.TargetPort.Type == intstr.Int {
```

Comment:
TargetPort can be omitted, and in that case, `mapping[svcportspec.Port] = svcportspec.Port` should be used

Reply:
Will fix that and add a test case for that.

Reply:
I think the only safe way to detect omitted is to assume the IntValue() is 0, but if there is a better way please let me know.

Reply:
hmm, looks like defaulting takes care of this for you and sets targetPort to match port when omitted (https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/v1/defaults.go#L120). Can you verify that is the case?